NULL check missing in agent instrumentation code

When the agent attempts to view the existing correlation headers it does not perform a NULL check [here](https://github.com/Microsoft/ApplicationInsights-Java/blob/36214d5c28eb975e008c5b7c60c4840fdd10ec20/agent/src/main/java/com/microsoft/applicationinsights/agent/internal/agent/HttpClientMethodVisitor.java#L66).

This can throw an exception during the filter pipeline and lead to an internal server error.

It is recommended that we add a check similar to [this one](https://github.com/Microsoft/ApplicationInsights-Java/blob/36214d5c28eb975e008c5b7c60c4840fdd10ec20/agent/src/main/java/com/microsoft/applicationinsights/agent/internal/agent/HttpClientMethodVisitor.java#L143-L146) to avoid a nullref exception.

It was see recently on a struts application using thew new 2.0.0-BETA SDK and form based security with j_security_check.

I have also hit this issue.  Here are the repro steps:

1.) Follow these steps to setup AI for dependencies: https://docs.microsoft.com/en-us/azure/application-insights/app-insights-java-agent

2.) Use the ToDo List sample app which is a Java SpringBoot app that uses the Azure Document DB library and App Insights library.

3.) Follow the ReadMe's steps to build and run the app.  When attempting to run the app, you'll get the following exception:

Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.microsoft.azure.documentdb.DocumentClient]: Factory method 'documentClient' threw exception; nested exception is java.lang.NullPointerException
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189) ~[spring-beans-4.3.12.RELEASE.jar!/:4.3.12.RELEASE]
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588) ~[spring-beans-4.3.12.RELEASE.jar!/:4.3.12.RELEASE]
        ... 88 common frames omitted
Caused by: java.lang.NullPointerException: null
        at com.microsoft.applicationinsights.web.internal.correlation.TelemetryCorrelationUtils.retrieveCorrelationContext(TelemetryCorrelationUtils.java:140) ~[applicationinsights-web-2.0.0-BETA.jar!/:na

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## Version 2.0.0
+
+- Fix #506 Null Reference Check causing Null Pointer Exception in `TelemetryCorrelationUtils.java`
+
 ## Version 2.0.0-BETA
 - Updating various dependencies to latest version
 - Introducing public class CustomClassWriter in Agent to enable finding common super classes used for Agent instrumentation without loading it

--- a/web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TelemetryCorrelationUtils.java
+++ b/web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TelemetryCorrelationUtils.java
@@ -148,7 +148,7 @@ public class TelemetryCorrelationUtils {
 
 		//check if context is null - no correlation will happen
 		if (ThreadContext.getRequestTelemetryContext() == null) {
-			InternalLogger.INSTANCE.error("No correlation wil happen, Thread context is null");
+			InternalLogger.INSTANCE.warn("No correlation wil happen, Thread context is null");
 			return "";
 		}
 
@@ -273,7 +273,7 @@ public class TelemetryCorrelationUtils {
 
 		//check if context is null - no correlation will happen
 		if (ThreadContext.getRequestTelemetryContext() == null) {
-			InternalLogger.INSTANCE.error("No correlation will happen " +
+			InternalLogger.INSTANCE.warn("No correlation will happen " +
 					"Thread context is null while resolving Correlation Context");
 			return;
 		}

--- a/web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TelemetryCorrelationUtils.java
+++ b/web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TelemetryCorrelationUtils.java
@@ -111,7 +111,15 @@ public class TelemetryCorrelationUtils {
 	public static String generateChildDependencyId() {
 		
 		try {
+
 			RequestTelemetryContext context = ThreadContext.getRequestTelemetryContext();
+
+			//check if context is null - no correlation will happen
+			if (context == null) {
+				InternalLogger.INSTANCE.warn("No Correlation will happen, Thread context is null while generating child dependency");
+				return "";
+			}
+
 			RequestTelemetry requestTelemetry = context.getHttpRequestTelemetry();
 
 			String parentId = requestTelemetry.getContext().getOperation().getParentId();
@@ -137,6 +145,13 @@ public class TelemetryCorrelationUtils {
 	 * @return The correlation context as a string.
 	 */
 	public static String retrieveCorrelationContext() {
+
+		//check if context is null - no correlation will happen
+		if (ThreadContext.getRequestTelemetryContext() == null) {
+			InternalLogger.INSTANCE.error("No correlation wil happen, Thread context is null");
+			return "";
+		}
+
 		CorrelationContext context = ThreadContext.getRequestTelemetryContext().getCorrelationContext();
 		return context.toString();
 	}
@@ -253,6 +268,13 @@ public class TelemetryCorrelationUtils {
 
 		if (baggages == null) {
 			InternalLogger.INSTANCE.warn("Could not access header information: " + CORRELATION_CONTEXT_HEADER_NAME);
+			return;
+		}
+
+		//check if context is null - no correlation will happen
+		if (ThreadContext.getRequestTelemetryContext() == null) {
+			InternalLogger.INSTANCE.error("No correlation will happen " +
+					"Thread context is null while resolving Correlation Context");
 			return;
 		}
 

